### PR TITLE
MCO-573: Wire up security/trust/pull secrets between rpm-ostree and internal registry

### PR DESCRIPTION
**Pull request by djoshy**
- Added a new service account for doing image pulls from the internal registry
- Added a new field in ControllerConfig for the secret generated from this new SA
- The secret is then merged with the global pull secret before being applied to the machine config, which templates it out to a file.
- The daemon now first attempts to link rpm-ostree's auth file to the new merged secret file(as long as it exists and not empty, otherwise it defaults to global pull secret). 
- Added a rebootless carveout for the "merged" secrets file so we do not reboot if the secret changes

This does not work for scaling up nodes - this only allows access for nodes that are already in the cluster.  [MCO-578](https://issues.redhat.com/browse/MCO-578) is intended for that. I did add a default route path in the image registry secret file, but solving the scale up problem needs some additional certs finesse. 

I've split out the commits for easier review, but I can squash if necessary. A lot of this was based on John's work from #3397. Thanks a bunch, @jkyros! 

How to test:
(i) [Load](https://docs.openshift.com/container-platform/4.13/registry/accessing-the-registry.html#registry-accessing-directly_accessing-the-registry) an RHCOS image into the MCO's slice of the image registry via podman/docker. 
(ii) Apply an MC to an infra node with an overriden osImageURL that points to this new image in the image registry
(iii) The infra node should successfully rebase to the new image(observe logs on the node)
(iv) The global pull secret should still work; this can be tested by rolling the node back to the original MC.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -46,6 +46,7 @@ require (
 	k8s.io/component-base v0.27.2
 	k8s.io/kubectl v0.27.2
 	k8s.io/kubelet v0.27.2
+	k8s.io/kubernetes v1.27.3
```
Comment (sinnykumari) [inline go.mod:49]: not mandatory but for consistency and avoid any surprise maybe we should also update other kube deps to match 1.27.3

**Comment by sinnykumari** [comment]
This looks to me that we need qe merge testing?

**Comment by djoshy** [comment]
> This looks to me that we need qe merge testing?

Yep, agreed. I'll mark it on jira (:

File: go.mod (hunk 1 above)
Comment (djoshy) [reply go.mod:49]: done!

**Comment by sergiordlr** [comment]
Verification steps:

1. Initiate a debug session in a node
` oc debug  node/$(oc get nodes -l node-role.kubernetes.io/master -ojsonpath="{.items[0].metadata.name}")`

2. Login to the internal api address

`oc login -u kubeadmin -p  ......   https://api-int.sregidor.qe.devcluster.openshift.com:6443`

3. Login to the internal registry

`podman login -u kubeadmin -p $(oc whoami -t) image-registry.openshift-image-registry.svc:5000`

4. Get the Base image

```
sh-5.1# oc adm release info --image-for "rhel-coreos" 2> /dev/null
registry.build05.ci.openshift.org/ci-ln-7s99r4k/stable@sha256:f88540a94d2915828ebe167b8e1ed8221322bd8f0525b3b96da6ec931d81fa1c
```

5. Create a test Dokerfile using the base image

```
sh-5.1# cat Dockerfile 
FROM registry.build05.ci.openshift.org/ci-ln-7s99r4k/stable@sha256:f88540a94d2915828ebe167b8e1ed8221322bd8f0525b3b96da6ec931d81fa1c
RUN echo 'Hello' >  /etc/hello-world-file
```

6. Build a new osImage using the dockerfile

```
sh-5.1# podman build -r image-registry.openshift-image-registry.svc:5000/openshift-machine-config-operator/new-os-image:latest .
...
STEP 2/2: RUN echo 'Hello' >  /etc/hello-world-file
COMMIT image-registry.openshift-image-registry.svc:5000/openshift-machine-config-operator/new-os-image:latest
--> 5057971daae
Successfully tagged image-registry.openshift-image-registry.svc:5000/openshift-machine-config-operator/new-os-image:latest
5057971daae558adae3ebb674058f291c26dc03e1baaec6da3e21a05cacb8e3f


sh-5.1# podman push image-registry.openshift-image-registry.svc:5000/openshift-machine-config-operator/new-os-image:latest
...
Writing manifest to image destination
Storing signatures

# Now we can see a new imagestream in the openshift-machine-config-operator namespace
 oc get is
NAME           IMAGE REPOSITORY                                                                                  TAGS     UPDATED
new-os-image   image-registry.openshift-image-registry.svc:5000/openshift-machine-config-operator/new-os-image   latest   37 seconds ago
```

we can get the digested image from the imagestream

```
$ oc get is -o yaml |grep sha256
        dockerImageReference: image-registry.openshift-image-registry.svc:5000/openshift-machine-config-operator/new-os-image@sha256:aac11fac56e2bc2d59d8eeec4137f5fd470c40c4f340efb0c15c7ea003c53c30

```
7. Create a new MC using the new osImage

```
kind: MachineConfig
apiVersion: machineconfiguration.openshift.io/v1
metadata:
  labels:
    machineconfiguration.openshift.io/role: "worker"
  name: auto-generated-osimageurl
spec:
  osImageURL: "image-registry.openshift-image-registry.svc:5000/openshift-machine-config-operator/new-os-image@sha256:aac11fac56e2bc2d59d8eeec4137f5fd470c40c4f340efb0c15c7ea003c53c30"

```
8. The new images should be deployed


We can see this in the MCD logs

```
I0811 11:24:35.985411    1365 rpm-ostree.go:262] Linking ostree authfile to /etc/mco/internal-registry-pull-secret.json

$ oc debug  node/$(oc get nodes -l node-role.kubernetes.io/worker -ojsonpath="{.items[0].metadata.name}")  -- chroot /host cat /etc/hello-world-file
Starting pod/ip-10-0-26-52us-east-2computeinternal-debug-zrmq8 ...
To use host binaries, run `chroot /host`
Hello

Removing debug pod ...
```


9. Deploy the usbguard extension


No problems found

```
sh-5.1# rpm-ostree status
State: idle
Deployments:
* ostree-unverified-registry:image-registry.openshift-image-registry.svc:5000/openshift-machine-config-operator/new-os-image@sha256:aac11fac56e2bc2d59d8eeec4137f5fd470c40c4f340efb0c15c7ea003c53c30
                   Digest: sha256:aac11fac56e2bc2d59d8eeec4137f5fd470c40c4f340efb0c15c7ea003c53c30
                  Version: 414.92.202308102017-0 (2023-08-11T13:53:55Z)
          LayeredPackages: usbguard

```


9. Remove the MC

The default osImage is re-applied without problems

10. Run tests
"[sig-mco] MCO Author:mhanss-NonPreRelease-High-42680-change pull secret in the openshift-config namespace [Serial]"
"[sig-mco] MCO Author:sregidor-NonPreRelease-Medium-57595-Use empty pull-secret[Disruptive] [Serial]"
"[sig-mco] MCO Layering Author:sregidor-ConnectedOnly-VMonly-Longduration-NonPreRelease-Critical-54085-Update osImage changing /etc /usr and rpm [Disruptive] [Serial]"
"[sig-mco] MCO Layering Author:sregidor-VMonly-ConnectedOnly-Longduration-NonPreRelease-High-54909-Configure extensions while using a custom osImage [Disruptive] [Serial]"

Tests passed

11. Verify that the difference between the pull-secret and the value stored in the ControllerConfig is

```
    "172.30.160.193:5000": {
    "default-route-openshift-image-registry.apps.sregidor.qe.devcluster.openshift.com": {
    "image-registry.openshift-image-registry.svc.cluster.local:5000": {
    "image-registry.openshift-image-registry.svc:5000": {
```


12. Modify the pull-secret

When we modify the pull-secret, the changes are not synced with the ControllerConfig

```
alias getpullsecret='function __lgb() { unset -f __lgb; set -x; oc get secret pull-secret -n openshift-config --template="{{index .data \".dockerconfigjson\" | base64decode}}" | jq; set +x; }; __lgb'
alias setpullsecret='function __lgb() { unset -f __lgb; set -x; oc set data secret pull-secret -n openshift-config --from-file=.dockerconfigjson=$1; set +x; }; __lgb'

getpullsecret > /tmp/pull-secret.json

# edit the pull secret to add a new fake section 
,
    "my.fake.repo": {
      "auth": "myfakeauth=="
    }

# Add update the pull secret in the cluster with our new change
setpullsecret  /tmp/pull-secret.json

```

We can see that the pull secret is updated in /var/lib/kubelet/config.json file, and we can find the fake section there.
But when we look for this information in the controllerconfig, there is only information regarding the internalregistry, the rest of the pull secret has been removed.

This was the list of repositoris in the controllerconfig before we updated the pull-secret

```
  "auths": {
    "brew.registry.redhat.io": {
    "registry.build05.ci.openshift.org": {
    "quay.io": {
    "cloud.openshift.com": {
    "registry.svc.ci.openshift.org": {
    "172.30.160.193:5000": {
    "default-route-openshift-image-registry.apps.sregidor.qe.devcluster.openshift.com": {
    "image-registry.openshift-image-registry.svc.cluster.local:5000": {
    "image-registry.openshift-image-registry.svc:5000": {
    "registry.ci.openshift.org": {
    "registry.redhat.io": {
    "registry.connect.redhat.com": {
    "registry.stage.redhat.io": {
    "brewregistry.stage.redhat.io": {
```


After updating the pull secret we have this

```
$ oc get controllerconfig machine-config-controller -ojsonpath='{.spec.internalRegistryPullSecret'} | base64 -d | jq |grep ": {"  
  "auths": {
    "172.30.160.193:5000": {
    "default-route-openshift-image-registry.apps.sregidor.qe.devcluster.openshift.com": {
    "image-registry.openshift-image-registry.svc.cluster.local:5000": {
    "image-registry.openshift-image-registry.svc:5000": {

```
All the information not related to the internal registry has been removed.

**Summary**
Currently the behavior in step 12 is being investigated. We can't add the approved label until it is finished.

**Comment by djoshy** [comment]
Thanks for the feedback @sergiordlr! I found that I wasn't reporting an error during the merge. The actual error log once I fixed that, turned out to be: `unable to parse auth field, must be formatted as base64(username:password)`

For the unmarshal to work correctly, every secret being added needs to be in the correct format. I copied in one of my other secrets and used that as the "new" dummy secret, and it is being rendered correctly. The global pull secret( `/var/lib/kubelet/config.json`) was propogated correctly because it isn't being unmarshalled; it is just directly templated out. Since we are reporting an unmarshal error now in case of incorrect format, I think this behaviour is ok.

**Comment by rioliu-rh** [comment]
@djoshy I think the Sergio's concern is: even the auth info in secret is dummy data, it is cannot be synced to controller config silently. the user does not have a chance to know that. I think if we can raise this data format issue in any condition of controller config that's better.  @sinnykumari WDYT

**Comment by djoshy** [comment]
@rioliu-rh Thanks for the feedback. Since the operator is the one putting the secrets together, the easiest way to surface this is via the operator conditions. I've made those changes , here is what they look like:
Operator condition:
```
$ oc get co/machine-config
NAME             VERSION                         AVAILABLE   PROGRESSING   DEGRADED   SINCE   MESSAGE
machine-config   4.14.0-0.ci-2023-08-11-124640   False       False         True       5m17s   Cluster not available for [{operator 4.14.0-0.ci-2023-08-11-124640}]: failed to merge global pull secret:   unmarshal failure: unable to parse auth field, must be formatted as base64(username:password)
```
Logs:
```
E0814 18:07:06.525714       1 sync.go:389] Merging registry secrets failed with: failed to merge global pull secret:   unmarshal failure: unable to parse auth field, must be formatted as base64(username:password)
I0814 18:07:06.716356       1 event.go:298] Event(v1.ObjectReference{Kind:"", Namespace:"openshift-machine-config-operator", Name:"machine-config", UID:"66cc7384-5166-419a-8cab-bdbf0dd832da", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Warning' reason: 'OperatorDegraded: RenderConfigFailed' Failed to resync 4.14.0-0.ci-2023-08-11-124640 because: failed to merge global pull secret:   unmarshal failure: unable to parse auth field, must be formatted as base64(username:password)
I0814 18:07:07.116153       1 event.go:298] Event(v1.ObjectReference{Kind:"", Namespace:"openshift-machine-config-operator", Name:"machine-config", UID:"66cc7384-5166-419a-8cab-bdbf0dd832da", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Warning' reason: 'RenderConfigFailed' Cluster not available for [{operator 4.14.0-0.ci-2023-08-11-124640}]: failed to merge global pull secret:   unmarshal failure: unable to parse auth field, must be formatted as base64(username:password)
```
Let me know if that works better!

**Comment by sinnykumari** [comment]
@djoshy  I think the concern here is existing pull secret data that was present in controllerconfig is replaced by just internal registry content when global pull secret was updated in step 12. Expected behavior here would be that controllerconfig should always have updated global pull secret + internal registry content ?

**Comment by djoshy** [comment]
Correct, the expectation is that the merged secret field will have both. However, if a pull secret of incorrect format is added, it will fail the merge process(unmarshalling into a docker config object won't work). And the failure will be propagated as shown in my previous comment. When Sergio was testing, I wasn't catching this error and it would go through silently, skipping over the pull secret. This should no longer be the case and we should always be propagating an error(and preserving the old pull secret)

**Comment by sinnykumari** [comment]
Thanks David for providing the explanation and making necessary changes. This looks good to me since we are now propogating error when updated pull secret is invalid and also controlleconfig preserves existing pull secret + internal registry related secret in case of error.
@rioliu-rh should be good for further testing

**Comment by rioliu-rh** [comment]
verified fake registry auth info in pull secret.
```console
cv
NAME      VERSION                                                   AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.14.0-0.ci.test-2023-08-15-130228-ci-ln-kk59xfk-latest   True        False         17m     Cluster version is 4.14.0-0.ci.test-2023-08-15-130228-ci-ln-kk59xfk-latest
```
get secret and update fake auth info
```console
oc get secret/pull-secret -n openshift-config --template='{{index .data ".dockerconfigjson" | base64decode}}' | jq . > /tmp/pull-secret.json

tail -5 /tmp/pull-secret.json
    "my.fake.repo": {
      "auth": "myfakeauth=="
    }
  }
}

oc set data secret pull-secret -n openshift-config --from-file=.dockerconfigjson=/tmp/pull-secret.json
secret/pull-secret data updated
```
check out status of co/machine-config 
```console
co machine-config -o yaml | yq -y '.status.conditions'
- lastTransitionTime: '2023-08-15T13:18:52Z'
  message: Error while reconciling 4.14.0-0.ci.test-2023-08-15-130228-ci-ln-kk59xfk-latest
  status: 'False'
  type: Progressing
- lastTransitionTime: '2023-08-15T13:58:41Z'
  message: 'Failed to resync 4.14.0-0.ci.test-2023-08-15-130228-ci-ln-kk59xfk-latest
    because: failed to merge global pull secret:   unmarshal failure: unable to parse
    auth field, must be formatted as base64(username:password)'
  reason: RenderConfigFailed
  status: 'True'
  type: Degraded
- lastTransitionTime: '2023-08-15T13:58:41Z'
  message: 'Cluster not available for [{operator 4.14.0-0.ci.test-2023-08-15-130228-ci-ln-kk59xfk-latest}]:
    failed to merge global pull secret:   unmarshal failure: unable to parse auth
    field, must be formatted as base64(username:password)'
  reason: RenderConfigFailed
  status: 'False'
  type: Available
- lastTransitionTime: '2023-08-15T13:59:55Z'
  reason: AsExpected
  status: 'True'
  type: Upgradeable

co machine-config
NAME             VERSION                                                   AVAILABLE   PROGRESSING   DEGRADED   SINCE   MESSAGE
machine-config   4.14.0-0.ci.test-2023-08-15-130228-ci-ln-kk59xfk-latest   False       False         True       6m17s   Cluster not available for [{operator 4.14.0-0.ci.test-2023-08-15-130228-ci-ln-kk59xfk-latest}]: failed to merge global pull secret:   unmarshal failure: unable to parse auth field, must be formatted as base64(username:password)
```
condition `Available` = `False`, condition `Degraded` = `True` with same reason `RenderConfigFailed`

/label qe-approved

**Comment by djoshy** [comment]
Made one more change to not report an error(we just log to the error stream instead) and degrade for the lack of the machine-os-puller service account. This was causing issue on the upgrade test since the service accounts may not be synced on the first render config sync of the operator. If there is a better approach to this, I'm open to it!

@rioliu-rh Not sure if QE is required for this change since it is minor but let me know (:

**Review by sinnykumari — COMMENTED** [review]
This overall looks good.
One question: since limitation was noted that it won't work for scale-up node, is internal registry secret gets added by default or it is gated behind something? or the scale-up issue  is mainly about pulling up osImageURL from internal registry won't work for scale-up node?

**Comment by sinnykumari** [comment]
Also, let's run payload test to avoid any unknown regression :)
/payload 4.14 nightly blocking

**Comment by djoshy** [comment]
> This overall looks good. One question: since limitation was noted that it won't work for scale-up node, is internal registry secret gets added by default or it is gated behind something? or the scale-up issue is mainly about pulling up osImageURL from internal registry won't work for scale-up node?

The issue with scale-up node is that the node comes up outside the cluster and won't have access to the internal pod service network. We'll need to expose a route to the registry so that the node can pull from there. Then there's also the issue of image registry certs as well, which need to be available on scale-up(this was done by a daemonset that image registry folks maintained). That bit may be solved by Charlie's work in https://issues.redhat.com/browse/MCO-552  though! There's some more background about this on John's original PR #3397

**Comment by sinnykumari** [comment]
some of the payload test failed, can you double check and make sure they are not this PR related?

**Pull request merged**